capsules/src/test/aes.rs

```diff
@@ -0,0 +1,237 @@
+
+pub struct Test<'a, A: 'a> {
+    aes: &'a A,
+
+    key: TakeCell<'a, [u8]>,
+    iv: TakeCell<'a, [u8]>,
+    source: TakeCell<'a, [u8]>,
+    data: TakeCell<'a, [u8]>,
+
+    encrypting: Cell<bool>,
+    mode_ctr: Cell<bool>,
+    use_source: Cell<bool>,
+}
+
+pub const DATA_OFFSET: usize = AES128_BLOCK_SIZE;
+const DATA_LEN: usize = 4 * AES128_BLOCK_SIZE;
+
+impl<'a, A: AES128<'a> + AES128Ctr + AES128CBC> Test<'a, A> {
+    pub fn new(aes: &'a A,
+               key: &'a mut [u8],
+               iv: &'a mut [u8],
+               source: &'a mut [u8],
+               data: &'a mut [u8])
+               -> Self {
+        Test {
+            aes: aes,
+
+            key: TakeCell::new(key),
+            iv: TakeCell::new(iv),
+            source: TakeCell::new(source),
+            data: TakeCell::new(data),
+
+            encrypting: Cell::new(true),
+            mode_ctr: Cell::new(true),
+            use_source: Cell::new(true),
+        }
+    }
+
+    pub fn run(&self) {
+        self.aes.enable();
+
+        // Copy key into key buffer and configure it in the hardware
+        self.key.map(|key| {
+            for (i, b) in KEY.iter().enumerate() {
+                key[i] = *b;
+            }
+
+            assert!(self.aes.set_key(key) == ReturnCode::SUCCESS);
+        });
+
+        // Copy mode-appropriate IV into IV buffer and configure it in the hardware
+        self.iv.map(|iv| {
+            let iv_mode = if self.mode_ctr.get() {
+                &IV_CTR
+            } else {
+                &IV_CBC
+            };
+            for (i, b) in iv_mode.iter().enumerate() {
+                iv[i] = *b;
+            }
+
+            assert!(self.aes.set_iv(iv) == ReturnCode::SUCCESS);
+        });
+
+        // Copy mode-appropriate source into source buffer
+        let source_mode = if self.encrypting.get() {
+            &PTXT
+        } else {
+            if self.mode_ctr.get() {
+                &CTXT_CTR
+            } else {
+                &CTXT_CBC
+            }
+        };
+        self.source.map(|source| for (i, b) in source_mode.iter().enumerate() {
+            source[i] = *b;
+        });
+
+        if !self.use_source.get() {
+            // Copy source into dest for in-place encryption
+            self.source.map_or_else(|| panic!("No source"), |source| {
+                self.data.map_or_else(|| panic!("No data"),
+                                      |data| for (i, b) in source.iter().enumerate() {
+                                          data[DATA_OFFSET + i] = *b;
+                                      });
+            });
+        }
+
+        if self.mode_ctr.get() {
+            self.aes.set_mode_aes128ctr(self.encrypting.get());
+        } else {
+            self.aes.set_mode_aes128cbc(self.encrypting.get());
+        }
+        self.aes.start_message();
+
+        let start = DATA_OFFSET;
+        let stop = DATA_OFFSET + DATA_LEN;
+
```

Comment:
Why start at `DATA_OFFSET` (16)?

Reply:
I believe that the point of the AES refactor is to allow the AES hardware to operate on arbitrary ranges of slices. Previously you could only encrypt/decrypt a mutable buffer from the beginning, because if you had sliced into it and passed it to the driver, you would never get the original slice back. In other words, whenever you wanted to encrypt only part of a buffer, you would need a copy just because of the HIL's restrictiveness. There are also some new options for encrypting/decrypting different parts of buffers: for example if the source isn't present, then in-place encryption should be done directly in the destination buffer.

As for why exactly it's 16, I imagine that's just an arbitrary offset for testing purposes, but @daniel-scs might know better.

Reply:
Ok, makes sense. I was a little bit concerned about this on nrf51dk which only has 16kB RAM but it seems to fit. You can check my PR